Add staking docs (#8988)
solana-labs/solana: docs/src/cli/delegate-stake.md

## Create a Stake Account

To delegate stake, you will need to move your tokens into a stake account.
All commands in this section assume you have
[set a default fee-payer](transfer-tokens.md#set-a-default-fee-payer).

Create a stake account with the `solana create-stake-account` command:

```bash
solana create-stake-account --from=<KEYPAIR> <ACCOUNT_KEYPAIR> <AMOUNT>
```

`<AMOUNT>` tokens are transferred from the account at `<KEYPAIR>` to a new
stake account at the public key of `<ACCOUNT_KEYPAIR>`.  Record the account
public key, as it will be used later to perform actions on the stake account.
The keypair file can be discarded. To authorize additional actions, you will
use `<KEYPAIR>`, not `<ACCOUNT_KEYPAIR>`.

### Advanced: Derive Stake Account Addresses

When we delegate stake, we delegate all tokens to a single validator. To
delegate to multiple validators we'll need multiple stake accounts. Creating
a new keypair for each account and managing those addresses can be cumbersome.
Fortunately, we can derive stake addresses using the `--seed` option:

```bash
solana create-stake-account --from=<KEYPAIR> <KEYPAIR> --seed=<STRING> <AMOUNT>
```

> `<STRING>` is arbitrary but typically a number corresponding to which derived

Author: fixed

> `<STRING>` is arbitrary but typically a number corresponding to which derived

Reviewer: I know we don't want to get into the nitty gritty RE UTF-8, but we will be in markets with alphabets where 1 char != 1 byte
```
$ export WINK32="😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉😉"
$ echo -n $WINK32 | wc --chars
32
$ echo -n $WINK32 | wc --bytes
128
```
Maybe it'd be a better idea to introduce `create-address-with-seed` first?  It'll fail on invalid input

> `<STRING>` is arbitrary but typically a number corresponding to which derived

Author: 😉

## Set Stake and Withdraw Authorities

Staking commands look to keypairs to authorize certain stake account
operations. It uses a stake authority to authorize stake delegation,
deactivating stake, splitting stake, and setting a new stake authority.  It
uses a withdraw authority to authorize withdrawing stake, and setting either
a new stake or withdraw authority.

Stake and withdraw authorities can be set when creating account via the
`--stake-authority` and `--withdraw-authority` options, or afterward with
the `solana stake-authorize` command. For example, to set a new authorized
staker, run:

```bash
solana stake-authorize <ACCOUNT_KEYPAIR> --stake-authority=<KEYPAIR> --new-stake-authority=<PUBKEY>
```

This will use the existing stake authority `<KEYPAIR>` to authorize a new
stake authority `<PUBKEY>` on the stake account `<ACCOUNT_KEYPAIR>`.

## Delegate Stake

To delegate stake, run:

```bash
solana delegate-stake --stake-authority=<KEYPAIR> <ACCOUNT_PUBKEY> <VOTE_PUBKEY>
```

`<KEYPAIR>` authorizes the operation on the account with address
`<ACCOUNT_PUBKEY>`. The stake is delegated to the vote account with
public key `<VOTE_PUBKEY>`.

> 

Author: I considered it, but couldn't think of a reason why an entry-level user would care.

> 

Reviewer: We've had plenty of "Why is my stake amount less than the account balance?" while the stake was warming up during TdS

> 

Reviewer 2: I tend to agree with Trent. I expect that once their are rewards to be had, there will be plenty of "Why am I not seeing rewards accumulate?" while stake is warming up.

> 

Author: I'm not opposed to it, but I'd prefer to put it in a follow-up PR and have those details reviewed separately. I'm not confident I have it all straight.

## Deactivate Stake

Once delegated, you can deactivate stake with the `solana deactivate-stake`
command:

```bash
solana deactivate-stake --stake-authority=<KEYPAIR> <ACCOUNT_PUBKEY>
```

`<KEYPAIR>` authorizes the operation on the account with address
`<ACCOUNT_PUBKEY>`.

Note that stake takes several epochs to "cool down". Attempts to delegate
stake in the cooldown period will fail.

## Split Stake

All stake in a stake account is delegated to a single validator. If you'd
prefer to spread your stake over multiple validators use the
`solana split-stake` command to duplicate stake accounts and split your
tokens between them:

```bash
solana split-stake --stake-authority=<KEYPAIR> <ACCOUNT_PUBKEY> <NEW_ACCOUNT_KEYPAIR> <AMOUNT>
```

`<ACCOUNT_PUBKEY>` is the existing stake account, `<KEYPAIR>` is the
authorized staker, `<NEW_ACCOUNT_KEYPAIR>` is the keypair for the new account,
and `<AMOUNT>` is the number of tokens to transfer to the new account.

## Withdraw Stake

> Transfer tokens out of a stake account with the `solana withdraw-stake` command:

Reviewer: This section should mention that withdraws are subject to both cool down and lockup

Author: Do you know the details of that, by chance? I'd have to go sift through the code.

Reviewer: Withdraw eligibility seems to be
- `0` if lockup is in force
- `0` if in warm up or fully staked
- `cooled_amount` if in cool down
- `balance` or `balance - rent_exempt` if undelegated, but initialized account
- `balance` if the account is uninitialized
Should be a piece of cake to distill that down! :upside_down_face:

Author: Okay, cool. I'm going to do warmup/cooldown details in a followup PR. This might require another page.

Reviewer: We should probably look into refactoring the code such that we can reuse part of it to calculate an "eligible for withdraw" value for display in `stake-account`

# Earn Staking Rewards

> After you have [recieved SOL](transfer-tokens.md), you might consider putting

Reviewer 3: s/recieved/received

it to use by delegating stake to a validator node. Solana is a Proof of Stake
blockchain, which means that holding tokens gives you the option to influence
which validator nodes participate in securing the ledger. The more tokens you

> hold, the more influence you have. The running assumption of any Proof of Stake

Reviewer 2: This isn't just an assumption, it is a fact. The running assumption is that *because* you have more to lose, you are more likely to behave legitimately.

Author: the tokens you hold may not be *your* tokens

Reviewer 2: I think this would be better: "The more tokens you hold, the more influence you have, because you have more to lose if a validator node acts maliciously."

Reviewer 3: In this paragraph we talk about "validator nodes" and in the following paragraph, they are "validators".  I think we can drop the word "node" as it doesn't add much here.  As the validator ecosystem continues to evolve, it's possible that single validator entities/people/companies could operate multiple nodes, or some other ambiguity, but validator companies could attract potential stakers to them without needing to describe their "node" or architecture necessarily.

blockchain is that the more tokens you hold, the more you have to lose if a
validator node acts maliciously.

As a tokenholder, it is your option to use your tokens as *stake*. Stake is

> what we call tokens when they reside in a *stake account*. From the stake

Reviewer 2: "As a tokenholder, it is your option to use your tokens as stake."
What is the sentence meant to communicate that isn't already said in the first paragraph?

"Stake is what we call tokens when they reside in a *stake account*."
Not crazy about this sentence. How about "Tokens start acting as stake when they are moved into a *stake account*"

Author: Rewrote that paragraph

account, you can delegate that stake to validators. Solana weights validator
votes by the amount of stake delegated to them, which gives those validators
more influence in determining then next valid block of transactions in the
blockchain. Solana then generates new SOL periodically to reward stakers and
validators. You earn more rewards the more stake you delegate.

## Create a Stake Account

> To delegate stake, you will need to move your tokens into a stake account.

Reviewer 3: s/move your tokens/move some or all of your tokens/

All commands in this section assume you have

> [set a default fee-payer](transfer-tokens.md#set-a-default-fee-payer).

Reviewer 3: Until you rip out `--keypair` entirely, this is not just the fee payer, but also the default staking and withdrawing authorities.  In the default fee payer page, the language is rather cavalier about having a fraction of a SOL sitting on a hot key, when that by our defaults presented below, that hot key has a lot of authority as well.

Author: Ripped it out

Create a stake account with the `solana create-stake-account` command:

```bash
solana create-stake-account --from=<KEYPAIR> <ACCOUNT_KEYPAIR> <AMOUNT>
```

`<AMOUNT>` tokens are transferred from the account at `<KEYPAIR>` to a new

> stake account at the public key of `<ACCOUNT_KEYPAIR>`.  Record the account

Reviewer 2: Do we need to hand-hold through or at least reference the need to create this new keypair?

Reviewer 3: +1 for adding a step to create the new keypair.

public key, as it will be used later to perform actions on the stake account.
The keypair file can then be discarded. To authorize additional actions, you
will use `<KEYPAIR>`, not `<ACCOUNT_KEYPAIR>`.

### Advanced: Derive Stake Account Addresses

> When we delegate stake, we delegate all tokens in the stake account to a single

Reviewer 2: Why switch from "you" to "we" for this paragraph?

> validator. To delegate to multiple validators we will need multiple stake

Reviewer 2: Comma after "To delegate to multiple validators"

string. To see what stake address the command will derive, use
`solana create-address-with-seed`:

```bash
solana create-address-with-seed --from=<PUBKEY> <SEED_STRING> STAKE
```

It will output a derived address, which can be used for the
`<ACCOUNT_PUBKEY>` argument in staking operations.

## Set Stake and Withdraw Authorities

Staking commands look to keypairs to authorize certain stake account
operations. They use the stake authority to authorize stake delegation,
deactivating stake, splitting stake, and setting a new stake authority.  They
use the withdraw authority to authorize withdrawing stake, and when setting either
a new stake or withdraw authority.

Stake and withdraw authorities can be set when creating account via the
`--stake-authority` and `--withdraw-authority` options, or afterward with
the `solana stake-authorize` command. For example, to set a new authorized
staker, run:

```bash
solana stake-authorize <ACCOUNT_KEYPAIR> --stake-authority=<KEYPAIR> --new-stake-authority=<PUBKEY>
```

This will use the existing stake authority `<KEYPAIR>` to authorize a new
stake authority `<PUBKEY>` on the stake account `<ACCOUNT_KEYPAIR>`.

> 

Reviewer 3: `s/<ACCOUNT_PUBKEY>/<STAKE_ACCOUNT_ADDRESS>/g` please.  This may be inconsistent with how the CLI presents it currently, but I think this language is easier to follow.  There's a lot of usage of that arg name in this page.

Reviewer: I'll :+1: this.  I think we should use "address" anywhere we're talking about the bit string used to recall an account.  That it occasionally happens to be a pubkey is merely an implementation detail

Author: Agreed, great suggestion, thanks! https://github.com/solana-labs/solana/pull/8998

## Delegate Stake

To delegate stake, run:

```bash
solana delegate-stake --stake-authority=<KEYPAIR> <ACCOUNT_PUBKEY> <VOTE_PUBKEY>
```

`<KEYPAIR>` authorizes the operation on the account with address
`<ACCOUNT_PUBKEY>`. The stake is delegated to the vote account with

> public key `<VOTE_PUBKEY>`.

Reviewer 3: "Each validator owns one (or more?) vote account.  By delegating the stake to that vote account's pubkey in the  `<VOTE_PUBKEY>` argument, you are delegating stake to whichever validator owns that account."  

Not sure if we want to touch on here the inevitable question "How do I find a validator to stake and how do I choose the best one for me?".  Probably deserves its own section "Choosing a Validator" where we don't give advice on how to choose, but just note that they can use/find tools to examine validators' performance and fees and make an informed decision.  Not sure if we can/should even say that much to keep our influence out of the network.

Author: I'll can add a "Choose a Validator" doc later.

Author: And added a few words about looking up a vote account address via `solana validators`.  Pretty weak without a "Choose a Validator" doc.

## Deactivate Stake

> Once delegated, you can deactivate stake with the `solana deactivate-stake`

Reviewer 3: s/Once delegated/To stop delegating your stake once it has been delegated/

Just to make it clear that `dectivate` is the opposite of `delegate`, which is not obvious, particularly next to `withdraw` which sounds like similar behavior.  Can we just re-name it `undelegate`?

command:

```bash
solana deactivate-stake --stake-authority=<KEYPAIR> <ACCOUNT_PUBKEY>
```

`<KEYPAIR>` authorizes the operation on the account with address
`<ACCOUNT_PUBKEY>`.

Note that stake takes several epochs to "cool down". Attempts to delegate
stake in the cool down period will fail.

## Split Stake

All stake in a stake account is delegated to a single validator. If you

> would prefer to spread your stake over multiple validators use the

Reviewer 2: This seems to imply that the only way to stake multiple validators is to create one stake account and then split it, which in fact, a user could just create a separate account instead.
I understand that you're trying to build a complete walkthrough, but maybe qualify this in some way?
"If you would prefer to spread these tokens between multiple validators, you can use..." would be better, I think.

Author: Great point.  Really, only need to `split-stake` when a tokens are not eligible for withdraw.

After you have [received SOL](transfer-tokens.md), you might consider putting
it to use by delegating *stake* to a validator. Stake is what we call tokens
in a *stake account*. Solana weights validator votes by the amount of stake
delegated to them, which gives those validators more influence in determining
then next valid block of transactions in the blockchain. Solana then generates
new SOL periodically to reward stakers and validators. You earn more rewards
the more stake you delegate.

## Create a Stake Account

To delegate stake, you will need to transfer some tokens into a stake account.
Create a stake account with the `solana create-stake-account` command:

```bash
solana-keygen new --no-passphrase -o stake-account.json
solana create-stake-account --from=<KEYPAIR> stake-account.json <AMOUNT> --stake-authority=<KEYPAIR> --withdraw-authority=<KEYPAIR>
```

`<AMOUNT>` tokens are transferred from the account at `<KEYPAIR>` to a new
stake account at the public key of stake-account.json.  Record the account
public key, as it will be used later to perform actions on the stake account.
The stake-account.json file can then be discarded. To authorize additional

> actions, you will use `<KEYPAIR>`, not the keypair at stake-account.json.

Reviewer: ```suggestion
public key, as it will be used later as the `<STAKE_ACCOUNT_ADDRESS>`
when performing actions on the stake account.  The stake-account.json file
can then be discarded. To authorize additional actions, you will use
`<KEYPAIR>`, not the keypair at stake-account.json.
```

Author: But the call to `create-stake-account` was above

Author: @t-nelson, rewrote it

Reviewer: Ah! Looks like my attempt to uphold line wrap broken the fine grained diff highlighting.  My goal was to link the throwaway keypair to the `<STAKE_ACCOUNT_ADDRESS>`.  I don't believe we make a smooth transition there

All commands in this section assume you have
[set a default fee-payer](transfer-tokens.md#set-a-default-fee-payer).

Create a stake account with the `solana create-stake-account` command:

```bash
solana-keygen new --no-passphrase -o stake-account.json
solana create-stake-account --from=<KEYPAIR> stake-account.json <AMOUNT> --stake-authority=<KEYPAIR> --withdraw-authority=<KEYPAIR>
```

`<AMOUNT>` tokens are transferred from the account at `<KEYPAIR>` to a new
stake account at the public key of stake-account.json.  Record the account
public key, as it will be used later to perform actions on the stake account.
The stake-account.json file can then be discarded. To authorize additional
actions, you will use `<KEYPAIR>`, not the keypair at stake-account.json.

### Set Stake and Withdraw Authorities

Staking commands look to keypairs to authorize certain stake account
operations. They use the stake authority to authorize stake delegation,
deactivating stake, splitting stake, and setting a new stake authority.  They
use the withdraw authority to authorize withdrawing stake, and when setting
either a new stake or withdraw authority.

> Stake and withdraw authorities can be set when creating account via the

Reviewer: ```suggestion
Stake and withdraw authorities can be set when creating an account via the
```

the `--seed` option:

```bash
solana create-stake-account --from=<KEYPAIR> <STAKE_ACCOUNT_KEYPAIR> --seed=<STRING> <AMOUNT> --stake-authority=<PUBKEY> --withdraw-authority=<PUBKEY>
```

`<STRING>` is an arbitrary string up to 32 characters, but will typically be a
number corresponding to which derived account this is. The first account might
be "0", then "1", and so on. The keypair used for `<STAKE_ACCOUNT_KEYPAIR>`

> acts as base key. The command derives a new address from the base address and

Reviewer: ```suggestion
be "0", then "1", and so on. The pubkey of `<STAKE_ACCOUNT_KEYPAIR>` acts as
the base address. The command derives a new address from the base address and
```

## Delegate Stake

> To delegate your stake to a validator, you will need to its vote account

Reviewer: ```suggestion
To delegate your stake to a validator, you will need to know its vote account
```

in the cool down period will fail.

## Withdraw Stake

Transfer tokens out of a stake account with the `solana withdraw-stake` command:

```bash
solana withdraw-stake --withdraw-authority=<KEYPAIR> <STAKE_ACCOUNT_ADDRESS> <RECIPIENT_ADDRESS> <AMOUNT>
```

`<STAKE_ACCOUNT_ADDRESS>` is the existing stake account, `<KEYPAIR>` is the
withdraw authority, and `<AMOUNT>` is the number of tokens to transfer to
`<RECIPIENT_ADDRESS>`.

## Split Stake

You may want to delegate stake to additional validators while your existing
stake is not eligible for withdrawal. It might not be eligible because it is
currently staked, cooling down, or locked up. To transfer tokens from an
existing stake account to a new one, use the `solana split-stake` command:

```bash
solana split-stake --stake-authority=<KEYPAIR> <STAKE_ACCOUNT_ADDRESS> <NEW_STAKE_ACCOUNT_KEYPAIR> <AMOUNT>
```

`<STAKE_ACCOUNT_ADDRESS>` is the existing stake account, `<KEYPAIR>` is the
stake authority, `<NEW_STAKE_ACCOUNT_KEYPAIR>` is the keypair for the new account,
and `<AMOUNT>` is the number of tokens to transfer to the new account.

To split a stake account into a derived account address, use the `--seed`

> option in the same way as with `solana create-stake-account`.

Reviewer: ```suggestion
option.  See
[Derive Stake Account Addresses](#advanced-derive-stake-account-addresses)
for details.
```
